optuna/progress_bar.py

```diff
@@ -0,0 +1,88 @@
+import logging
+from typing import Any
+from typing import Optional
+
+from tqdm.auto import tqdm
+
+from optuna import logging as optuna_logging
+
+_tqdm_handler = None  # type: Optional[TqdmLoggingHandler]
+
+
+# Reference: https://gist.github.com/hvy/8b80c2cedf02b15c24f85d1fa17ebe02
+class TqdmLoggingHandler(logging.StreamHandler):
+
+    def emit(self, record: Any) -> None:
+        try:
+            msg = self.format(record)
+            tqdm.write(msg)
+            self.flush()
+        except (KeyboardInterrupt, SystemExit):
+            raise
+        except Exception:
+            self.handleError(record)
+
+
+class _ProgressBar(object):
+    """Progress Bar implementation for `Study.optimize` on the top of `tqdm`.
+
+    Args:
+        is_valid:
+            Whether to show progress bars in `Study.optimize`.
+        n_trials:
+            The number of trials.
+        timeout:
+            Stop study after the given number of second(s).
+    """
+
+    def __init__(
+        self,
+        is_valid: bool,
+        n_trials: Optional[int] = None,
+        timeout: Optional[float] = None,
+    ) -> None:
+        self._is_valid = is_valid
+        self._n_trials = n_trials
+        self._timeout = timeout
+
+        if self._is_valid:
+            self._progress_bar = tqdm(range(n_trials) if n_trials is not None else None)
+            global _tqdm_handler
+
+            tqdm_handler = TqdmLoggingHandler()
```

Comment:
You're declaring a local handler here (and using it in the following couple of lines) but you probably meant the global handler.